baseplate/experiments/providers/__init__.py

```diff
@@ -103,20 +119,25 @@ def parse_experiment(config):
         override = config.get("global_override")
         return ForcedVariantExperiment(override)
 
-    if experiment_type == "r2":
-        return R2Experiment.from_dict(
+    experiment_class = type_class_map.get(experiment_type, ForcedVariantExperiment)
+
+    if experiment_type in ['r2', 'feature_flag']:
+        return experiment_class.from_dict(
             id=experiment_id,
             name=name,
             owner=owner,
             version=version,
             config=experiment_config,
         )
-    elif experiment_type == "feature_flag":
-        return FeatureFlag.from_dict(
+    elif experiment_type in ['single_variant', 'multi_variant',
+                             'feature_rollout']:
+        return experiment_class.from_dict(
             id=experiment_id,
             name=name,
             owner=owner,
-            version=version,
+            start_ts=start_ts,
+            stop_ts=stop_ts,
+            enabled=enabled,
```

Comment:
Why pass this stuff to the experiment?  This stuff should be handled by `ForcedVariantExperiment`-s rather than the other providers so you don't have to worry about it.

Reply:
Setting aside how it was done in the old types, I disagree for a few reasons:
1) an experiment being disabled (or not started, or ended) doesn't make it a fundamentally different experiment. It just makes it disabled so that it doesn't bucket users. 
2) if we don't pass this information in, then we can't use it in the experiment. I can see cases where we want to enforce different restrictions based on the type of experiment. For instance, multivariant may be required to run for 2 weeks, whereas a multi-armed bandit may be required to run longer. 
3) If we don't pass this in, it makes it more complex to allow us to start an experiment in the future (or end one). Assume the following flow:
a) I create an experiment, set to start in an hour
b) a bucketing call is made, but the experiment hasn't started yet, so it gets created as a ForcedVariantExperiment. This experiment is then added to the experiment_cache
c) No further config changes happen for a day

Now it's going to be treated as a forced variant until the config has been re-parsed for some reason.

Reply:
I still think that the current way this is handled is the way to go, not implementing it in the experiment provider:

1. While true, now you have to implement this logic in all of your experiment providers, the reason for putting this in the parsing logic and using ForcedVariantExperiments is so you don't have to re-implement the same "disable" logic every time.
2. But we _don't_ need to use this right now, if we have some new type of experiment where we do need to use it in the future, we can pass it in/implement it there.
3. I'm not sure where an experiment would be cached for an hour, the only cache is local to the request, it's not shared between requests.

Reply:
1) but that's fundamentally a part of the experiment. I'd be willing to agree that it could be in the base experiment class (I put it in the base class for a simple experiment for that reason), but the status, start, and end times of an experiment are features of an experiment, not features of an experiment parser.
2) You're right, in that the hypothetical situations are just that: hypothetical. I still think that the properties of an experiment (whether it's enabled, when it's meant to start and end) should be _available_ to the experiment, not used to infer the type of experiment (that's what the `experiment_type` is for)
3) I may be showing my ignorance on some things here, but is there anything inherent in the `baseplate.Experiments` class that restricts its scope to request level? When you request bucketing on an experiment that hasn't been seen yet, its config is parsed, and the experiment object is added to this dict: https://github.com/reddit/baseplate/blob/master/baseplate/experiments/__init__.py#L67
That's the cache I'm referring to. 

Generally speaking though, I think using an entirely new type of object for the sake of abstracting away the enabled/disabled/started/ended has more downsides than upsides.

Reply:
I'm not sure which chain to respond to here but I wanted to hop in. I'm seeing a lot of discussion of "override in subclasses" and "implement in base classes". In general, I'd really strongly like to avoid 1) implementation inheritance, and 2) deep hierarchies of objects.  This ends up being quite byzantine to follow (jumping back and forth between class definitions to follow a code path) and very brittle.  Rather, it'd be good to encapsulate specific pieces of logic in functions or discrete objects we bring together with composition. Duplication by putting into multiple classes is definitely a bummer, but so is putting it into a base class instead. Would it make sense to put this kind of shared logic into an object that only handles expiration logic and then it can be included in any experiment type that cares?

Reply:
ah! I see @pacejackson had a much better suggested structure for this here: https://github.com/reddit/baseplate/pull/186/files#r198285273 :)

Reply:
As far as (3) goes, the `Experiments` objects are created by the `ExperimentsContextFactory` when you get a new request/context so you get a new cache on each request.